### PR TITLE
Implement Globus Auth with simple frontend for user profiles

### LINKED ISSUE #169: Implement Globus Auth
# Story
_As a ServiceX user I want to authenticate with Globus (or some other federated identity provider) so I can use my institutional identity_

# Assumptions
1. Globus Auth will enable users to login with their identity from a number of supported institutional providers (CERN and most universities). We will identify users by their Globus ID, and store this in the PostgreSQL database.
2. Each user will be issued a JWT refresh token by ServiceX, which will be stored in the database and visible on the web frontend.
3. Users will store their ServiceX refresh token a local `.servicex` file. The Python frontend client will use this to obtain access tokens and use these for authorization.
4. If a user leaks their refresh token, we can enable them to generate a new one on the website. Only the current refresh token can be used to obtain access tokens, so the old refresh token is no longer usable.

# Acceptance Criteria
1. Globus Auth is implemented
2. Web frontend allows users to sign in, create/view/edit their user profiles, and obtain their ServiceX tokens. At least one of the following is supported: copy the refresh token to clipboard, download a `.servicex` file with the token pre-populated, or copy and run a script which stores the refresh token in `~/.servicex` automatically
3. Python frontend supports reading refresh tokens from `.servicex` config file. The refresh token is used to obtain an access token which the client uses to submit transformation requests.
4. The Flask app has an additional endpoint which enables authenticated users to click a button and generate a new refresh token, which is stored in the database. When an access token is requested, Flask checks the refresh token against the current one in the database, and only issues an access token if there is a match.

### PR DESCRIPTION
**Pull request by AndrewEckart**
The scope of this PR is currently as follows:

- Authenticate users via Globus Auth
- If we have no user in the db with the corresponding Globus id, ask the user to create a profile
- Users who are authenticated and saved to the db can view/edit their profiles
- Generate ServiceX API tokens (JWT refresh tokens) for each user
- Enable users to view, copy, and invalidate/regenerate their tokens on the web
- Update API resources to use new tokens
- Clean up / remove old registration and authentication code
- Add unit tests

Additional items to be optionally implemented later:
- Button to download a `.servicex` file containing the current API token
- Bash script to create a `.servicex` file containing the current API token in the user's home directory

Closes ssl-hep/ServiceX#169

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -45,19 +45,24 @@
     include_package_data=True,
     zip_safe=False,
     install_requires=[
-        'Flask==1.1.2',
```
Comment (BenGalewsky) [inline setup.py:48]: Why delete all of the library version tags? It's important to pin the libraries down so a Docker build doesn't break functionality.

Comment (AndrewEckart) [reply setup.py:48]: How is this going to work with the new GitLab flow? It seems like the library versions should only ever be frozen in release branches, not in develop. If we fix them all now and only ever update them manually, then our dependencies will slowly become outdated.

File: servicex/models.py [hunk 2]
```diff
@@ -40,13 +40,14 @@ class UserModel(db.Model):
...
-    full_name = db.Column(db.String(120), nullable=False)
+    experiment = db.Column(db.String(120))
+    id = db.Column(db.Integer, primary_key=True)
     institution = db.Column(db.String(120))
-    key = db.Column(db.String(120), nullable=False)
+    name = db.Column(db.String(120), nullable=False)
     pending = db.Column(db.Boolean, default=True)
-    experiment = db.Column(db.String(120))
+    refresh_token = db.Column(db.Text, nullable=False, unique=True)
+    sub = db.Column(db.String(120), nullable=False, unique=True, index=True)
```
Comment (BenGalewsky) [inline servicex/models.py:50]: I know this name makes sense in the JWT world, but maybe it should be renamed to `subject` or `jwt_subject`, or just plain `identity`

File: migrations/versions/b389abb05262_v1_0_rc_1.py [hunk 3]
```diff
@@ -47,12 +47,13 @@ def upgrade():
...
-    sa.Column('full_name', sa.String(length=120), nullable=False),
+    sa.Column('experiment', sa.String(length=120), nullable=True),
     sa.Column('institution', sa.String(length=120), nullable=True),
-    sa.Column('key', sa.String(length=120), nullable=False),
+    sa.Column('name', sa.String(length=120), nullable=False),
     sa.Column('pending', sa.Boolean(), nullable=True),
-    sa.Column('experiment', sa.String(length=120), nullable=True),
+    sa.Column('refresh_token', sa.Text, nullable=False, unique=True),
     sa.Column('updated_at', sa.DateTime(), nullable=True),
+    sa.Column('sub', sa.String(120), nullable=False, unique=True, index=True),
```
Comment (BenGalewsky) [inline migrations/versions/b389abb05262_v1_0_rc_1.py]: This is messed up, there should be a v1.0.0.0-RC3 database version.. don't try to change released db schema.

File: servicex/web/create_profile.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            sub = session['sub']
+            new_user = UserModel(
+                sub=sub,
+                email=form.email.data,
+                name=form.name.data,
+                institution=form.institution.data,
+                experiment=form.experiment.data,
+                refresh_token=create_refresh_token(identity=sub))
+            try:
+                new_user.save_to_db()
```
Comment (BenGalewsky) [inline servicex/web/create_profile.py:34]: What about duplicate user profiles. Should we check before saving?

File: servicex/web/slack_interaction.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        try:
+            UserModel.accept(email)
+        except NoResultFound as err:
+            respond(response_url, user_not_found(str(err)))
+            return Response(status=404)
+        response_msg = signup_ia(original_msg, initiating_user, action_id)
+        slack_response = requests.post(response_url, response_msg)
+        slack_response.raise_for_status()
+    elif action_id == "reject_user":
+        # todo blocked by PR for delete-user endpoint
```
Comment (BenGalewsky) [inline servicex/web/slack_interaction.py]: This us unblocked, no?

File: migrations/versions/b389abb05262_v1_0_rc_1.py (hunk 3 above)
Comment (AndrewEckart) [reply migrations/versions/b389abb05262_v1_0_rc_1.py]: Didn't really know what it was but I'll take a look at the Alembic docs

File: servicex/models.py (hunk 2 above)
Comment (AndrewEckart) [reply servicex/models.py:50]: Seems like it's really an OAuth term.. apparently used by OAuth providers who adhere to the OpenID Connect standard:

>The OpenID Connect ID Token is a signed JSON Web Token (JWT) that is given to the client application along side the regular OAuth access token. The ID Token contains a set of claims about the authentication session, including an identifier for the user (sub), the identifier for the identity provider who issued the token (iss), and the identifier of the client for which this token was created (aud).

I'll rename it to `oauth_subject` (I don't want to confuse it with `id` which is our ServiceX user ID, an integer). If we wanted to support providers other than Globus down the road and those providers also complied with OIDC (as CILogon does) it probably wouldn't be too bad.

File: migrations/versions/b389abb05262_v1_0_rc_1.py (hunk 3 above)
Comment (AndrewEckart) [reply migrations/versions/b389abb05262_v1_0_rc_1.py]: Okay, this should be fixed now.. sorry about that.

**Comment by AndrewEckart** [comment]
Seems there is still a problem with my migration code:
```
INFO  [alembic.runtime.migration] Running upgrade  -> b389abb05262, V1.0-RC.1
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1278, in _execute_context
    cursor, statement, parameters, context
  File "/usr/local/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 593, in do_execute
    cursor.execute(statement, parameters)
psycopg2.errors.DuplicateTable: relation "requests" already exists
```

**Review by BenGalewsky — APPROVED** [review]
This is a great achievement. Glad to be able to merge this!

**Pull request merged**